File: c++/verticalmove.cpp
#include "verticalmove.h"
#include "color.h"
#include "piece.h"
#include "game.h"
#include "pawn.h"
#include "queen.h"
#include "illegalmoveexception.h"
#include <math.h>

namespace Model
{
    VerticalMove::VerticalMove(Game *game, Piece *piece, const Square &startSquare, const Square &endSquare)
    {
        this->game = game;
        this->piece = piece;
        //capturedPiece = game->getPiece(endSquare);
        pawnPromotedTo = NULL;
        this->startSquare = startSquare;
        this->endSquare = endSquare;
        capturedPiece = NULL;
        wasMoved = piece->getWasMoved();
    }

    void VerticalMove::doMove()/* throws IllegalMoveException */
    {
        if (capturedPiece != NULL)
            game->removePieceFromList(capturedPiece);
        game->setPieceOnChessboard(endSquare, piece);
        game->setPieceOnChessboard(startSquare, NULL);
        piece->setCurrentSquare(endSquare);
        piece->setWasMoved(true);
        game->changeTurn();
        if (game->isKingInCheck() == true) {
            undoMove();
            throw IllegalMoveException();
        }
        if (dynamic_cast<Pawn *>(piece))
            if (piece->getColor() == WHITE)
            {
                if (endSquare.getRank() == 7)
                {
                    game->removePieceFromList(piece);
                    pawnPromotedTo = new Queen(game, WHITE, endSquare);
                    game->addPieceToList(pawnPromotedTo);
                    game->setPieceOnChessboard(endSquare, pawnPromotedTo);
                }
            }
            else if (endSquare.getRank() == 0)
            {
                game->removePieceFromList(piece);
                pawnPromotedTo = new Queen(game, WHITE, endSquare);
                game->addPieceToList(pawnPromotedTo);
                game->setPieceOnChessboard(endSquare, pawnPromotedTo);
            }
        //Logger.print(this, "The move has been done.");
    }

    bool VerticalMove::isForwardMove()
    {
        return startSquare.getRank() - endSquare.getRank() > 0 && piece->getColor() == WHITE
                || startSquare.getRank() - endSquare.getRank() < 0 && piece->getColor() == BLACK;
    }

    bool VerticalMove::isOneStepMove()
    {
        return abs(startSquare.getRank() - endSquare.getRank()) == 1;
    }

    bool VerticalMove::isTwoStepsMove()
    {
        return abs(endSquare.getRank() - startSquare.getRank()) == 2;
    }

    bool VerticalMove::isPathClear()
    {
        Square square = endSquare;
        if (game->isSquareFree(square))
            square = nextSquare(square);
        else if (piece->getColor() != game->getPiece(square)->getColor())
        {
            capturedPiece = game->getPiece(square);
            square = nextSquare(square);
            //Logger.print(this, "Capture..");
        }
        else {
            //Logger.print(this, "Path isn't clear.");
            return false;
        }
        while (square != startSquare)
        {
            if (!game->isSquareFree(square))
            {
                //Logger.print(this, "Path isn't clear.");
                return false;
            }
            square = nextSquare(square);
        }
        //Logger.print(this, "The path is clear...");*/
        return true;
    }

    Square VerticalMove::nextSquare(const Square &currentSquare)
    {
        Square nextSquare;
        nextSquare.setFile(currentSquare.getFile());
        if (endSquare.getRank() < startSquare.getRank())
            nextSquare.setRank(currentSquare.getRank() + 1);
        else
            nextSquare.setRank(currentSquare.getRank() - 1);
        return nextSquare;
    }

    void VerticalMove::undoMove()
    {
        game->setPieceOnChessboard(startSquare, piece);
        game->setPieceOnChessboard(endSquare, capturedPiece);
        if (capturedPiece != NULL)
            game->addPieceToList(capturedPiece);
        if (pawnPromotedTo != NULL) {
            game->removePieceFromList(pawnPromotedTo);
            game->addPieceToList(piece);
        }
        piece->setCurrentSquare(startSquare);
        piece->setWasMoved(wasMoved);
        game->changeTurn();
        //Logger.print(this, "The move has been undone.");
    }
}

File: java/src/model/VerticalMove.java
package model;

import logger.Logger;
import exceptions.IllegalMoveException;

/**
 * @author Jakub Rak
 */
class VerticalMove extends Move {
	private final Model model;
	private final Piece piece;
	private Piece capturedPiece;
	private Piece pawnPromotedTo;
	private final Square startSquare;
	private final Square endSquare;
	private final boolean wasMoved;

	VerticalMove(final Model model, final Piece piece, final Square startSquare, final Square endSquare) {
		this.model = model;
		this.piece = piece;
		capturedPiece = model.getPiece(endSquare);
		pawnPromotedTo = null;
		this.startSquare = startSquare;
		this.endSquare = endSquare;
		capturedPiece = null;
		wasMoved = piece.wasMoved();
	}

	@Override
	void doMove() throws IllegalMoveException {
		if (capturedPiece != null)
			model.removePieceFromList(capturedPiece);
		model.setPieceOnChessboard(endSquare, piece);
		model.setPieceOnChessboard(startSquare, null);
		piece.setCurrentSquare(endSquare);
		piece.setWasMoved(true);
		model.changeTurn();
		if (model.isKingInCheck() == true) {
			undoMove();
			throw new IllegalMoveException();
		}
		if (piece instanceof Pawn)
			if (piece.getColour() == Colour.WHITE) {
				if (endSquare.getRank() == 7) {
					model.removePieceFromList(piece);
					pawnPromotedTo = new Queen(model, Colour.WHITE, endSquare);
					model.addPieceToList(pawnPromotedTo);
					model.setPieceOnChessboard(endSquare, pawnPromotedTo);
				}
			}
			else if (endSquare.getRank() == 0) {
				model.removePieceFromList(piece);
				pawnPromotedTo = new Queen(model, Colour.WHITE, endSquare);
				model.addPieceToList(pawnPromotedTo);
				model.setPieceOnChessboard(endSquare, pawnPromotedTo);
			}
		Logger.print(this, "The move has been done.");
	}

	@Override
	Piece getCapturedPiece() {
		return capturedPiece;
	}

	@Override
	Square getEndSquare() {
		return endSquare;
	}

	@Override
	Piece getPiece() {
		return piece;
	}

	@Override
	Square getStartSquare() {
		return startSquare;
	}

	/**
	 * Checks if it is a forward move.
	 * @return true if it is a forward move, otherwise false
	 */
	boolean isForwardMove() {
		return startSquare.getRank() - endSquare.getRank() > 0 && piece.getColour() == Colour.BLACK
				|| startSquare.getRank() - endSquare.getRank() < 0 && piece.getColour() == Colour.WHITE;
	}

	/**
	 * Checks if it is one step move.
	 * @return true if it is one step move, otherwise false
	 */
	boolean isOneStepMove() {
		return Math.abs(startSquare.getRank() - endSquare.getRank()) == 1;
	}

	@Override
	boolean isPathClear() {
		Square square = endSquare;
		if (model.isSquareFree(square) == true)
			square = nextSquare(square);
		else if (piece.getColour() != model.getPiece(square).getColour()) {
			capturedPiece = model.getPiece(square);
			square = nextSquare(square);
			Logger.print(this, "Capture..");
		}
		else {
			Logger.print(this, "Path isn't clear.");
			return false;
		}
		while (square.equals(startSquare) == false) {
			if (model.isSquareFree(square) == false) {
				Logger.print(this, "Path isn't clear.");
				return false;
			}
			square = nextSquare(square);
		}
		Logger.print(this, "The path is clear...");
		return true;
	}

	/**
	 * Checks if it is two steps move.
	 * @return true if it is two steps move, otherwise false
	 */
	boolean isTwoStepsMove() {
		return Math.abs(endSquare.getRank() - startSquare.getRank()) == 2;
	}
	
	/**
	 * Gets next square on move's path (direction from {@link #endSquare} to {@link #startSquare})
	 * @param currentSquare - current square
	 * @return next square
	 */
	private Square nextSquare(final Square currentSquare) {
		final Square nextSquare = new Square();
		nextSquare.setFile(currentSquare.getFile());
		if (endSquare.getRank() < startSquare.getRank())
			nextSquare.setRank(currentSquare.getRank() + 1);
		else
			nextSquare.setRank(currentSquare.getRank() - 1);
		return nextSquare;
	}

	@Override
	void undoMove() {
		model.setPieceOnChessboard(startSquare, piece);
		model.setPieceOnChessboard(endSquare, capturedPiece);
		if (capturedPiece != null)
			model.addPieceToList(capturedPiece);
		if (pawnPromotedTo != null) {
			model.removePieceFromList(pawnPromotedTo);
			model.addPieceToList(piece);
		}
		piece.setCurrentSquare(startSquare);
		piece.setWasMoved(wasMoved);
		model.changeTurn();
		Logger.print(this, "The move has been undone.");
	}

	@Override
	boolean wasMoved() {
		return wasMoved;
	}
}
